### PR TITLE
Import TPAS Satisfaction data

### PR DESCRIPTION
**Pull request by dwhenry**
Raw(ish) data will be provided via daily email.

Matching row are grouped and then assigned an integer
indicating which 'row' they represent, this data is then
used to create a UID, allowing data to be imported
multiple times.

NOTE: This process only supports growing datasets,
if the number of rows decreases, this will NOT be
correctly updated.

### CODE REVIEW

File: lib/importers.rb [hunk 1]
```diff
@@ -2,5 +2,6 @@ module Importers
   autoload :Google, 'importers/google'
   autoload :SmartSurvey, 'importers/smart_survey'
   autoload :TP, 'importers/tp'
+  autoload :Tpas, 'importers/tpas'
```
Comment (benbarnett) [inline lib/importers.rb]: Hmm, TP and Tpas? Aren't they the same?

File: lib/importers/tpas/record.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
+require 'csv'
+
+module Importers
+  module Tpas
+    class Record
+      VALID_SATISFACTION_VALUES = '0'..'4'
```
Comment (benbarnett) [inline lib/importers/tpas/record.rb]: I'm not a rubyist but is this normal to do with strings?

Comment (benlovell) [inline lib/importers/tpas/record.rb]: Yes.

File: lib/importers.rb (hunk 1 above)
Comment (dwhenry) [inline lib/importers.rb]: Renamed to tpas_satisfaction

File: app/jobs/import_tpas_satisfaction_data_job.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+require 'importers'
+
+class ImportTpasSatisfactionDataJob < ActiveJob::Base
```
Comment (benlovell) [inline app/jobs/import_tpas_satisfaction_data_job.rb]: I know we've a few of these now, but the `Job` suffixes on these classes are bumming me out. They're already contextualised as 'jobs' via their containing folder. Might be one for the future though so don't let that block this PR.

File: lib/importers/tpas_satisfaction/importer.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      )
+        @retriever = retriever.new(config: config)
+        @record = record
+        @saver = saver
+      end
+
+      def import
+        @retriever.process_emails do |email|
+          records = @record.build(io: email.file)
+          records && @saver.new(records: records).save
```
Comment (benlovell) [inline lib/importers/tpas_satisfaction/importer.rb]: Feels like perhaps the `saver` should just noop if records are `nil` i.e. we push that logic down, so we could drop the boolean expression here.

File: lib/importers/tpas_satisfaction/record.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        @uid ||= begin
+          md5 = Digest::MD5.new
+          @cells.each { |value| md5 << value.to_s }
+          md5 << @duplicate_index.to_s
+          md5.hexdigest
+        end
+      end
+
+      def given_at
+        Time.zone.parse("#{@cells[1]} #{@cells[2]}")
```
Comment (benlovell) [inline lib/importers/tpas_satisfaction/record.rb]: Maybe we should provide readers for `1` and `2` as we've done elsewhere here?

File: lib/importers/tpas_satisfaction/retriever.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        @config = config
+        @mail_retriever = mail_retriever
+      end
+
+      def process_emails
+        retriever = @mail_retriever.new(config: @config)
+        emails = retriever.search(search_keys: @config.search_string, file_name_regexp: @config.file_name_regexp)
+
+        succeeded, failed = emails.partition { |email| yield email }
+        failed.each { |email| Rails.logger.warn("TPAS email failed to process: #{email.subject}") }
```
Comment (benlovell) [inline lib/importers/tpas_satisfaction/retriever.rb]: I can't remember what you did elsewhere, but perhaps for some consistency and future decoupling we should inject the `Rails.logger`?

File: lib/importers/tpas_satisfaction/saver.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      def initialize(records:)
+        @records = records.select(&:valid?)
+      end
+
+      def save
+        ActiveRecord::Base.transaction do
+          satisfaction!
+        end
+      rescue CodeLookup::MissingMappingError => e
+        Bugsnag.notify(e)
```
Comment (benlovell) [inline lib/importers/tpas_satisfaction/saver.rb]: Same with this `Bugsnag` dependency, perhaps we should inject as a matter of course?

File: spec/features/import_tpas_satisfaction_data_spec.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+  scenario 'Re-importing satisfaction data does not create duplicates' do
+    when_i_import_tpas_data
+    and_i_import_tpas_data_a_second_time
+    then_the_satisfaction_data_has_been_saved
+  end
+
+  def when_i_import_tpas_data
+    setup_imap_server(
+      attachment:  File.read(Rails.root.join('spec/fixtures/tpas_satisfaction.csv'), mode: 'rb')
```
Comment (benlovell) [inline spec/features/import_tpas_satisfaction_data_spec.rb]: Extra whitespace has crept into the keyword argument here.

File: spec/features/import_tpas_satisfaction_data_spec.rb [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    Importers::TpasSatisfaction::Importer.new.import
+  end
+
+  def setup_imap_server(attachment:)
+    mail_attachment = double( # Replace with local IMap server
+      :mail_attachment,
+      file: StringIO.new(attachment),
+      uid: SecureRandom.uuid
+    )
+    allow_any_instance_of(MailRetriever).to receive(:search).and_return([mail_attachment])
```
Comment (benlovell) [inline spec/features/import_tpas_satisfaction_data_spec.rb]: I cringe at `allow_any_instance_of`. It's presence hints at a smell, usually an opaque dependency. Can we inject a double for this somewhere instead?

File: app/jobs/import_tpas_satisfaction_data_job.rb (hunk 3 above)
Comment (dwhenry) [inline app/jobs/import_tpas_satisfaction_data_job.rb]: done

File: lib/importers/tpas_satisfaction/importer.rb (hunk 4 above)
Comment (dwhenry) [inline lib/importers/tpas_satisfaction/importer.rb]: Records can never be nil here. worst case is an empty array, so have removed boolean guard.

File: lib/importers/tpas_satisfaction/record.rb (hunk 5 above)
Comment (dwhenry) [inline lib/importers/tpas_satisfaction/record.rb]: 💯

File: lib/importers/tpas_satisfaction/retriever.rb (hunk 6 above)
Comment (dwhenry) [inline lib/importers/tpas_satisfaction/retriever.rb]: As discussed, we should think about this in the wide scope.

File: spec/features/import_tpas_satisfaction_data_spec.rb (hunk 9 above)
Comment (dwhenry) [inline spec/features/import_tpas_satisfaction_data_spec.rb]: Fixed this and the other two places it was done.

**Comment by benlovell** [comment]
:shipit:

**Pull request merged**